osu.Server.Spectator.Tests/Multiplayer/HostManagementTests.cs

```diff
@@ -53,5 +56,67 @@ public async Task HostLeavingCausesHostTransfer()
             using (var room = await Rooms.GetForUse(ROOM_ID))
                 Assert.True(room.Item?.Host?.UserID == USER_ID_2);
         }
+
+        [Fact]
+        public async Task HostKicksUser()
+        {
+            SetUserContext(ContextUser);
+            await Hub.JoinRoom(ROOM_ID);
+
+            SetUserContext(ContextUser2);
+            await Hub.JoinRoom(ROOM_ID);
+
+            var kickedUserReceiver = new Mock<IMultiplayerClient>();
+            Clients.Setup(clients => clients.Client(USER_ID_2.ToString())).Returns(kickedUserReceiver.Object);
+
+            SetUserContext(ContextUser);
+            await Hub.KickUser(USER_ID_2);
+
+            // other players received the event.
+            Receiver.Verify(r => r.UserKicked(It.Is<MultiplayerRoomUser>(u => u.UserID == USER_ID_2)), Times.Once);
+
+            // the kicked user received the event.
+            kickedUserReceiver.Verify(r => r.UserKicked(It.Is<MultiplayerRoomUser>(u => u.UserID == USER_ID_2)), Times.Once);
+
+            using (var room = await Rooms.GetForUse(ROOM_ID))
+                Assert.True(room.Item?.Users.All(u => u.UserID != USER_ID_2));
+
+            await Assert.ThrowsAsync<KeyNotFoundException>(() => UserStates.GetForUse(USER_ID_2));
+        }
+
+        [Fact]
+        public async Task HostAttemptsToKickSelf()
+        {
+            SetUserContext(ContextUser);
+            await Hub.JoinRoom(ROOM_ID);
+
+            SetUserContext(ContextUser2);
+            await Hub.JoinRoom(ROOM_ID);
+
+            SetUserContext(ContextUser);
+            await Assert.ThrowsAsync<InvalidStateException>(() => Hub.KickUser(USER_ID));
+
+            Receiver.Verify(r => r.UserKicked(It.IsAny<MultiplayerRoomUser>()), Times.Never);
+
+            using (var room = await Rooms.GetForUse(ROOM_ID))
+                Assert.True(room.Item?.Users.Any(u => u.UserID != USER_ID_2));
+        }
+
+        [Fact]
+        public async Task NonHostAttemptsToKickUser()
+        {
+            SetUserContext(ContextUser);
+            await Hub.JoinRoom(ROOM_ID);
+
+            SetUserContext(ContextUser2);
+            await Hub.JoinRoom(ROOM_ID);
+
+            await Assert.ThrowsAsync<NotHostException>(() => Hub.KickUser(USER_ID));
+
+            Receiver.Verify(r => r.UserKicked(It.IsAny<MultiplayerRoomUser>()), Times.Never);
+
+            using (var room = await Rooms.GetForUse(ROOM_ID))
+                Assert.True(room.Item?.Users.Any(u => u.UserID != USER_ID_2));
```

Comment:
This one is definitely wrong, though. `ContextUser2` was doing the kicking while not being host, so there was never any danger that they would disappear (in any sane scenario, that is). Should presumably read

```suggestion
                Assert.True(room.Item?.Users.Any(u => u.UserID == USER_ID));
```

too.

Reply:
These both have the same meaning with two users right? Definitely could read better but I believe it’s logically testing the same thing.

Reply:
Uhh right, those assertions are in fact equivalent iff there are 2 users. Not sure how it happened that I misread this... I think testing the desired outcome directly rather than its double negative may be safer/more understandable?

Reply:
yeah, i'll change it for readability